cdap-app-fabric/src/main/java/io/cdap/cdap/internal/capability/CapabilityApplier.java

```diff
@@ -0,0 +1,402 @@
+      String capability = capabilityConfig.getCapability();
+      CapabilityConfig existingConfig = capabilityReader.getConfig(capability);
+      if (capabilityConfig.equals(existingConfig)) {
+        capabilityWriter.deleteCapabilityOperation(capability);
+        continue;
+      }
+      capabilityWriter.addOrUpdateCapabilityOperation(capability, CapabilityAction.DISABLE, capabilityConfig);
+      LOG.debug("Disabling capability {}", capability);
+      capabilityWriter
+        .addOrUpdateCapability(capabilityConfig.getCapability(), CapabilityStatus.DISABLED, capabilityConfig);
+      //stop all the programs having capability metadata. Services will be stopped by SystemProgramManagementService
+      stopAllProgramsWithMetadata(capability);
+      capabilityWriter.deleteCapabilityOperation(capability);
+      LOG.debug("Disabled capability {}", capability);
+    }
+  }
+
+  private void deleteCapabilities(Set<CapabilityConfig> deleteSet) throws Exception {
+    for (CapabilityConfig capabilityConfig : deleteSet) {
+      String capability = capabilityConfig.getCapability();
+      CapabilityConfig existingConfig = capabilityReader.getConfig(capability);
+      //already deleted
+      if (existingConfig == null) {
+        capabilityWriter.deleteCapabilityOperation(capability);
+        continue;
+      }
+      capabilityWriter.addOrUpdateCapabilityOperation(capability, CapabilityAction.DELETE, capabilityConfig);
+      LOG.debug("Deleting capability {}", capability);
+      if (existingConfig.getStatus() == CapabilityStatus.ENABLED) {
+        //stop all the programs having capability metadata.
+        stopAllProgramsWithMetadata(capability);
+      }
+      //remove all applications having capability metadata.
+      deleteAllAppsWithMetadata(capability);
+      //remove deployments of system applications
+      for (SystemApplication application : capabilityConfig.getApplications()) {
+        ApplicationId applicationId = getApplicationId(application);
+        deleteAppWithRetry(applicationId);
+      }
+      capabilityWriter.deleteCapability(capability);
+      capabilityWriter.deleteCapabilityOperation(capability);
+      LOG.debug("Deleted capability {}", capability);
+    }
+  }
+
+  private ApplicationId getApplicationId(SystemApplication application) {
+    String version = application.getVersion() == null ? ApplicationId.DEFAULT_VERSION : application.getVersion();
+    return new ApplicationId(application.getNamespace(), application.getName(), version);
+  }
+
+  private ProgramId getProgramId(SystemProgram program) {
+    ApplicationId applicationId = new ApplicationId(program.getNamespace(), program.getApplication(),
+                                                    program.getVersion());
+    return new ProgramId(applicationId, ProgramType.valueOf(program.getType().toUpperCase()), program.getName());
+  }
+
+  private void deleteAllAppsWithMetadata(String capability) throws Exception {
```

Comment:
The method signature has nothing to do with metadata. Metadata is the implementation detail. You better call this `deleteAllAppsWithCapability`.

Reply:
Done